Various improvements

RedTheKitsune opened the pull request:
- Added support for setting entity to currently compiled blueprint by tracking opened blueprint editors:
I found that VS Code Wakatime plugin sets ``--entity`` to currently edited file so I decided to add this feature to this plugin.
The ``OnBlueprintPreCompile`` event provide UBlueprint class but it was triggering on every blueprint compile, so to fix this I subscribed to ``OnAssetOpened`` and ``OnAssetClosed`` to track what asset is opened and closed so its gonna send heartbeat only for the blueprint that we have opened.
- Redone wakatime-cli check to use unreal native method instead of cli command:
For some reason calling cmd command returns weird exit code, which causes redownloading wakatime-cli even if it already exists.
- Added option for setting custom ``api-url``
This is for all the people that use wakapi. I added the field to plugin window popup, but I think it shouldnt be there and code should grab it from config file.
- Blueprint compiling related heartbeat is now plugged in after editor window is initialized to prevent heartbeat spam during editor startup blueprint compilation
The pre editor window during startup is compiling all blueprints in the project, this caused a lot of heartbeats to be generated, so to fix it code hooks to blueprint events in the ``OnEditorInitialized``.
- Added option to specify Language for the heartbeat to differentiate between editor and blueprint heartbeat
This fixes #41 

Other changes:
Thanks to wakatime-cli log and docs I found out that the only thing to provide git branch info is to add ``--project-folder`` with path to the project so ``FPaths::ProjectDir()`` was enough to add it. Fixes #28 
I changed all hardcoded path ``/`` to ``\\`` becouse cmd has some ``/`` arguments and it was behaving badly with some paths. I also converted all runtime paths to it, but...
...at the end didnt change anything becouse now sending heartbeat softlocks editor so I changed it to powershell and it works now.

Experiments:
I tried wrapping heartbeat into AsyncTask but at that time it was behaving weirdly.
Tried changing ``CreateProcess`` to unreal native method for running external commands/processes but I also run into some issues. Tho it allows to easily check the return code and StdOut/StdErr and automatically waits.

Additional ideas:
I think it will be possible to introduce more tracking of what is edited by adding more asset type checks to ``OnAssetOpened``/``OnAssetClosed`` and hooking into some "OnAssetSave" event?

Im keeping it currently drafted. I want to add grabbing plugin version from uproject before it gets merged to master.

RedTheKitsune committed:
- Added support for setting entity to currently compiled blueprint by tracking opened blueprint editors

- Redone wakatime-cli check to use unreal native method instead of cli command
- Added option for setting custom api url
- Blueprint compiling related heartbeat is now plugged in after editor window is initialized to prevent heartbeat spam during editor startup blueprint compilation
- Added option to specify Language for the heartbeat to differentiate between editor and blueprint heartbeat

TheAshenWolf commented:
Thank you for the wonderfully written summary! All of these contributions are much appreciated!

Let me know if you have any questions, or if you feel like it's ready, so I can give it a check in Unreal :)

RedTheKitsune committed:
- Added code to read plugin version from .uplugin

- Changes to make api_url optional when saving

RedTheKitsune committed:
- Redone saving code to allow more flexibility

- Read config file again to get fresh config data, before showing settings window

RedTheKitsune committed:
- Changed TMap to std::map

- Fixed bIsDirty for config map

RedTheKitsune commented:
So I added code to automatically read plugin version from .uproject.
Redone the way code reads and updates keys in wakatime .ini file becouse it was getting really ugly when it had to save the values.
The settings window automatically updates itself based on config file, previously it was reading whatever it had in config during editor opening.

TheAshenWolf commented:
**Checking on UE 5.3.2**
First of all, the plugin didn't build by the simple "Click to rebuild". Then I figured out it was caused by me moving the project to another drive, so Unreal got a bit confused.

After fixing that, however, I encountered
`WakaTimeForUE.cpp(158): Error C2039 : 'OnAssetClosedInEditor': is not a member of 'UAssetEditorSubsystem'`
and 
`WakaTimeForUE.cpp(630): Error C2039 : 'OnAssetClosedInEditor': is not a member of 'UAssetEditorSubsystem'`

After commenting out these lines, I was able to start up the engine. Was this perhaps added in 5.4 or 5.5?
- [ ] Maybe a little `#if ENGINE_MAJOR_VERSION >= 5 && ENGINE_MINOR_VERSION >= 5` is in order.

I also noticed that the editor freezes for a second every time a heartbeat is sent; this may have always been the case or might be caused by me having the editor on an HDD, but since you didn't touch the code and I have seen nobody talk about it, I wouldn't worry about it too much.

Can confirm that running a CMD command instead of PS command freezes the editor.

WakaTime website does not show the heartbeats yet, but Rider did notices the categories (Rider submits only "Coding") 
![image](https://github.com/user-attachments/assets/03aefc85-f8d6-4661-bbdd-c75b3c2c2999)


Everything else looks great to me; would you mind patching that `OnAssetClosedInEditor` thing?

RedTheKitsune committed:
- Added preprocessor directives to fix UE versions <5.4 failing to compile

RedTheKitsune commented:
I wrapped entire BPs name tracking functionality into preprocessor directives, becouse that `OnAssetClosedInEditor` is necessary for knowing which BPs we have opened, so we can detect when opened BP is compiled in `OnBlueprintPreCompile`. Otherwise it will send bunch of BPs that we dont edit but are dependency of the BP that is currently compiled.

Tested on 5.4.4 and 5.3.2. The only difference is missing BP names in wakatime in 5.3.2 but thats intended.

Regarding heartbeat freezes, I could try wrapping it in AsyncTask so it will get processed and sent in other thread. If its not gonna work then I could add a option to `RunCommand` to ignore waiting for the command to finish.

RedTheKitsune committed:
- Add comment explaining preprocessor directive for <UE5.4
